Inner blocks: Reduce tree depth to improve performance.

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## Related PRs:

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/5746

## What?
<!-- In a few words, what is the PR actually doing? -->
Reduce the rendering complexity and depth of the block list for inner blocks.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617. 

Reducing rendering complexity is likely to improve performance overall. Also,
React Native has shown it has a limit in regards to number of nested views it
can render. Rendering too deep of trees can result in crashes; we observed this
when List v2 began using inner blocks.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- Combine `BlockList` and `BlockListCompact` to reduce code duplication between
  the separate components, e.g. item render, footer, end-of-list block appender
  button, empty list placeholder.
- Prepare support for additional block types to use the non-virtualized list
  through the addition of features like the nested block appender.

The following [diagram](https://excalidraw.com/#json=oXqmzQFp0lERueKKFVowL,RtdT1z4t0wFbkDG7yPbpXg) visually showcases the new structure of the block list used for 
both the root and inner blocks. 
![block-list-refactor](https://github.com/WordPress/gutenberg/assets/438664/03db4620-b02a-4fcb-9f2b-9aaa16d2ee28)


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
We should test adding, editing, re-ordering, and removing various block types,
including nested blocks. Additionally, we should test various device
orientations and sizes.

Specifically, we should heavily test various organizations of inner blocks with 
different alignments applied to the parent and children blocks. The testing 
instructions for a couple of Columns block PRs are a good start: https://github.com/WordPress/gutenberg/pull/19013, 
https://github.com/WordPress/gutenberg/pull/25621. 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->
n/a

dcalhoun committed:
test: Await overlooked asynchronous task

The test helper executes asynchronous updates to the layout. If this is
not awaited, test failures can occur in certain circumstances.

dcalhoun committed:
feat: Combine BlockList and BlockListCompact

Reduce code duplication between the separate components, e.g. item
render, footer, end-of-list block appender button, empty list
placeholder.

dcalhoun committed:
wip: Gallery, Columns working; Buttons broken

Partially fix multi-column layouts by passing existing styles to a
wrapping view. All of the passed styles may not be necessary, need to
investigate. Also, Buttons render broken, wrapping lines and overflowing
the parent container.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -328,29 +220,170 @@ export default function BlockList( {
-		<BlockListConsumer>
-			{ ( { scrollRef } ) =>
-				renderList( {
-					parentScrollRef: scrollRef,
```
@geriux is this ref still required if all inner blocks no longer render a virtualized list? 

From my review of the code, this appears to merely prevent an inner blocks' virtualized list from overriding its parent's ref stored in this file's `scrollViewRef`, but `parentScrollRef` is not actually used by child components. If that is accurate, passing this may no longer be necessary. WDYT?

geriux replied:
You're correct, since we are changing the way we render InnerBlocks, we won't need to do this anymore and we'll only need to set the main list ref `scrollRef` so it can be used by the Drag&Drop feature.

As it looks now, you took the right approach to remove the usage of `BlockListConsumer`.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -328,29 +220,170 @@ export default function BlockList( {
+				<>
+					{ blockClientIds.length > 0 ? (
+						<View
+							style={ [
+								getStyles(
+									isRootList,
+									isStackedHorizontally,
+									horizontalAlignment
+								),
+								horizontal && styles.horizontalContentContainer,
+								isWider( blockWidth, 'medium' ) &&
+									( isContentStretch && isMultiBlocks
+										? styles.horizontalContentContainerStretch
+										: styles.horizontalContentContainerCenter ),
+							] }
```
These styles were applied in bulk to "fix" multi-column inner blocks. It needs further review, as all of the styles may not be necessary for inner blocks. Also, the Buttons block is not properly displaying Buttons inline, they erroneously wrap and overflow the parent container.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -311,6 +202,7 @@ export default function BlockList( {
 			<BlockListItem
 				index={ index }
 				isStackedHorizontally={ isStackedHorizontally }
+				key={ clientId }
```
While `keyExtractor` will set this for a virtualized list, the simple, mapped list needs `key` explicitly set.

dcalhoun reviewed (commented):
👋🏻 @geriux. I published this draft to share the initial approach. This draft is very raw. I combined the components without much deduplication with a focus on getting a working version, with the plan to follow up to make it "right," then make it fast. It's easiest to review with "Hide whitespace" enabled. 

Also, I wanted to ask a question regarding the scroll view refs. Please see that in the inline comments.

dcalhoun commented:
@geriux also, if there aspects of https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617 or https://github.com/wordpress-mobile/gutenberg-mobile/issues/5618 to which you would like to begin contributing towards, feel free to do so and push to this branch as necessary. I believe we could resolve any merge conflicts.

dcalhoun committed:
test: Fix nested lists tests

Nested lists now rely upon `BlockListItem`, which returns `null` if the
`blockWidth` has not yet been set. In order for test queries for nested
list items to succeed, we must trigger the `onLayout` callback for the
nested block lists. `BlockListItem` is now used to expand capabilities
for nested blocks, e.g. multi-column grid items.

geriux commented:
> @geriux also, if there aspects of [wordpress-mobile/gutenberg-mobile#5617](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617) or [wordpress-mobile/gutenberg-mobile#5618](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5618) to which you would like to begin contributing towards, feel free to do so and push to this branch as necessary. I believe we could resolve any merge conflicts.

I can take https://github.com/wordpress-mobile/gutenberg-mobile/issues/5618 😎

dcalhoun committed:
wip: Fix Blocks button layout

Fixes Buttons, but Columns remain broken.

dcalhoun committed:
refactor: Remove unused BlockList title prop

dcalhoun committed:
refactor: Separate inner block list styles

The list is no longer shared, so we can merely set the appropriate
styles on each list element.

dcalhoun committed:
refactor: Remove unused virtualized listKey

Now that inner blocks do not use a virtualized list, a unique list key
is no longer necessary.

dcalhoun committed:
refactor: Remove outdated inline comment

The scroll ref is no longer passed to inner blocks as they do not use
virtualized lists.

dcalhoun committed:
wip: Explore nest Columns fixes

The relocated styles may need to be separate from the top-level block
list element, as the styles may conflict with other styles. It does not
fix Columns, however.

Commit:
Mobile - BlockList - Fix layout issues for stacked horizontally blocks

dcalhoun committed:
fix: Revert Buttons alignment workaround

This caused issues for non-Buttons inner block alignment. The original
issue of Buttons inner blocks not rendering inline was addressed in
43c0b14918f8ed03037c01d94321922dc31a7fa3.

dcalhoun committed:
fix: Column width sums less than 100% correct align

These styles caused columns to align to the center when their width sum
did not equal or execeed 100%. Removing these styles did not appear to
negatively impact other inner blocks or the use cases outlined in
https://github.com/WordPress/gutenberg/pull/25621.

dcalhoun committed:
refactor: Inner blocks use clientId key

This mirrors the approach for cells of top-level block lists. It is also
likely more robust, providing better performance when reordering blocks.

dcalhoun committed:
refactor: Remove unnecessary key

The inner blocks list now sets a key on a wrapping view, rather than via
the `renderItem` function. Thus, this key is no superfluous.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -74,7 +66,6 @@ export default function BlockList( {
 	renderAppender,
 	renderFooterAppender,
 	rootClientId,
-	title,
```
From what I could find, this prop is unused. The `header` prop is utilized for displaying the post title in the root block list.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -190,114 +181,6 @@ export default function BlockList( {
-					) }
-					data={ blockClientIds }
-					keyExtractor={ identity }
-					listKey={
-						rootClientId ? `list-${ rootClientId }` : 'list-root'
-					}
```
This is no longer necessary now that only the root list is virtualized.

dcalhoun commented on packages/block-editor/src/components/block-list/test/index.native.js:
```diff
@@ -52,7 +52,7 @@ describe( 'BlockList', () => {
 			const socialLinksBlock = await getBlock( screen, 'Social Icons' );
 			fireEvent.press( socialLinksBlock );
-			triggerBlockListLayout( socialLinksBlock );
+			await triggerBlockListLayout( socialLinksBlock );
```
We should always await the asynchronous `triggerBlockListLayout` helper. The lack of `await` was erroneous and can cause `act` warnings and test failures in certain circumstances.

dcalhoun commented on packages/block-library/src/list/test/edit.native.js:
```diff
@@ -70,6 +70,7 @@ describe( 'List block', () => {
 		// Select List block
 		const [ listBlock ] = screen.getAllByLabelText( /List Block\. Row 1/ );
 		fireEvent.press( listBlock );
+		await triggerBlockListLayout( listBlock );
```
Now that inner blocks rely upon `BlockListItem`, the nested list values are not rendered until the `onLayout` handler is triggered to set the `blockWidth`.

dcalhoun commented on packages/block-library/src/list/test/edit.native.js:
```diff
@@ -163,6 +166,12 @@ describe( 'List block', () => {
 		const indentButton = screen.getByLabelText( 'Indent' );
 		fireEvent.press( indentButton );
 
+		// Await recently indented list item layout
+		const [ listItemBlock1 ] = screen.getAllByLabelText(
+			/List item Block\. Row 1/
+		);
+		await triggerBlockListLayout( listItemBlock1 );
```
Resolves an `act` warning in the test logs.

dcalhoun committed:
fix: Prevent block list footer from stretching to parent height

The parent flex styles were applied to the footer element. This wrapping
view prevents those styles from erroneously stretching the provided
footer element.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -328,29 +211,159 @@ export default function BlockList( {
+		( isFloatingToolbarVisible ? floatingToolbar.height : 0 );
+
+	return (
+		<View
+			style={ containerStyle }
+			onAccessibilityEscape={ clearSelectedBlock }
+			onLayout={ onLayout }
+			testID="block-list-wrapper"
 		>
```
The one outstanding issue of which I am aware is that this `View` wrapping inner blocks causes an inability to consistently move the cursor into a List Item block rich text field. So far, I am not able to identify why this is the case.

geriux replied:
Hey! I've pushed this change https://github.com/WordPress/gutenberg/pull/50447/commits/b46922b357c60711d0864abe527efcd318294791 that sets the option to not render the appender. Without this, it was making the `EmptyList` to be rendered with the list items making the tap functionality to not work as expected.

dcalhoun reviewed (commented):
👋🏻 @geriux. I consider this PR ready for review and testing at this point. I created installable builds for [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/20660#issuecomment-1542740525) and [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/18397#issuecomment-1542931298). 

There is one known issue with lists that I noted in an inline comment.

Commit:
Mobile - List block - Remove usage of useCompactList and disables the option to render the appender for InnerBlocks

geriux commented:
I'm currently focusing on testing, so far on Android I see no regressions or performance issues with these new changes. I tested on both a Redmi Note 8T (Android 11) and a Samsung A23 5G (Android 13) ✅

geriux commented:
I finished testing on an iPhone 14 Pro (OS 16.4.1) and on an iPad (iPad OS 16.1), on both portrait and landscape including the mentioned column PRs testing cases and its working as expected ✅

geriux reviewed (commented):
Code changes look good to me! I think the only thing pending is to see why CI is failing over [Gutenberg mobile](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5746).

dcalhoun committed:
test: Update Android e2e Buttons block Xpath selector

This Xpath selector became outdated with the block list refactor.

dcalhoun committed:
test: Delay query for block actions menu

Appium reported this cached element had been removed from the DOM.
Relocating the query seemingly resolved this issue.

geriux reviewed (approved):
LGTM! Nice work @dcalhoun 👏👏!
